Remove the need to pass state and parentUiState to dashboard_panel directive

stacey-gammon opened the pull request:
This improves isolation of the panel and makes state changes clearer.

Along the way I simplified the logic for loading the saved object and once simplified, I didn't think it required a separate file so brought the logic inside the directive.

Also made note of where https://github.com/elastic/kibana/issues/9523 is creeping in.  I was originally going to fix it in this PR but since how to fix it isn't clear yet, this PR remains a non-functional change.

weltenwort commented on src/core_plugins/kibana/public/dashboard/directives/dashboard_panel.js:
```diff
@@ -69,52 +52,124 @@ uiModules
+         */
+        let postLoadInit;
+
+        if ($scope.panel.type === 'search') {
```
The problem I have with this explicit conditional is that it does not scale well to new panel types:
* Previously, when adding a new panel type, a new loader would just have to be created and added to the declarative `panelTypes` mapping in `loadPanelProvider`. The panel code itself would not have to be changed.
* This conditional though would grow with each new panel type and probably become a lot harder to read.

Stacey-Gammon replied:
Got rid of the conditional.  New types just need to register a loader in getObjectLoadersForDashboard.  No good place right now for custom post initialization code, but ideally there wouldn't need to be any.

Stacey-Gammon commented:
What do you think about this new version @weltenwort?  I extracted the load_saved_object code and removed the need for scope by pushing the "postInit" function stuff into initializePanel which is probably a better spot for it anyhow, since that is "postInit" code.

I created a new service for the loaders, so to add a new type all you have to do is add one to that service.  Any custom initialization of the saved object must, for now, go in the directive.

I'd like to improve that latter requirement, but I think this PR is sufficient and that can be a Step 3. Plus I don't have a clear idea right now how to improve the initialization code.

kobelb commented on src/core_plugins/kibana/public/dashboard/directives/dashboard_panel.js:
```diff
@@ -69,52 +57,89 @@ uiModules
+        // This causes changes to a saved search to be hidden, but also allows
+        // the user to locally modify and save changes to a saved search only in a dashboard.
+        // See https://github.com/elastic/kibana/issues/9523 for more details.
+        $scope.panel.columns = $scope.panel.columns || $scope.savedObj.columns;
```
This used to only happen for saved searches. Are $`scope.panel.columns` and `$scope.panel.sort` only being set so that we can watch them and know when to call `saveState`?

weltenwort replied:
True, maybe a `if ($scope.savedObj.type === 'search') { ... }` would be appropriate? And then line 97 could also read `if ($scope.savedObj.type === 'visualization')` for symmetry.

Stacey-Gammon replied:
Yea, I figured it was just a noop for visualizations, but I can wrap in an if as well.

Stacey-Gammon commented on src/core_plugins/kibana/public/dashboard/directives/dashboard_panel.js:
```diff
@@ -1,15 +1,19 @@
 import 'ui/visualize';
 import 'ui/doc_table';
-import { loadPanelProvider } from 'plugins/kibana/dashboard/components/panel/lib/load_panel';
+import 'plugins/kibana/dashboard/services/get_object_loaders_for_dashboard.factory';
```
This was motivated by some conversations with @w33ble and a new style pattern the Management team has been experimenting with.  This morning I decided I'm on the fence with whether or not I should be incorporating this new pattern, read more here: https://github.com/elastic/kibana/pull/9859#issuecomment-272880859

Once I hear back from @w33ble and @cjcenizal on that comment I'll decide whether to revert the `.factory` naming scheme.

FWIW I do like the pattern a lot but since it's not fully adopted by the team, it does seem a bit premature, especially with the last email that went around about consistency trumping new patterns until a certain # of criteria is hit.

Stacey-Gammon replied:
Here too, reverting this change to avoid introducing new patterns that aren't yet embraced by the team as a whole.

Stacey-Gammon commented on src/core_plugins/kibana/public/dashboard/directives/dashboard_panel.js:
```diff
@@ -1,15 +1,19 @@
-.directive('dashboardPanel', function (savedVisualizations, savedSearches, Notifier, Private, $injector) {
-  const loadPanel = Private(loadPanelProvider);
+.directive('dashboardPanel', function ($injector) {
+  const Private = $injector.get('Private');
```
another new style suggestion from the Management team.  Once you are only using the injector, you don't need to wrap the class in Private, so it'll make getting rid of Private easier.

weltenwort replied:
Well, `FilterManagerProvider` is still wrapped below. Isn't the only reason you get around using `Private` on `getObjectLoadersForDashboard` the fact that it has been globally registered under that name? That is still tightly coupled to Angular's DI, then.

Stacey-Gammon replied:
Yea, I think the idea is to stop using nameless functions that require Private, and explicitly register them in angular.  The more I think about this specific change though, the more I am unsure.

I was originally under the assumption that if I got rid of the parameters and only passed $Injector, I could avoid wrapping `dashboardPanel` with `Private` but that doesn't make sense because dashboardPanel is a directive, first off, and second off, is registered with angular, and no class registered with angular would require Private.

After going back to the original discussion with @w33ble I think the idea is to use $injector only for nameless functions, not for things already registered with angular.  Does that sound right, @w33ble?  So this should go back to it's previous version?

w33ble replied:
> That is still tightly coupled to Angular's DI

Everything will be, you can't use Angular without the DI 😢 .

There's actually 2 things going on here though. First is the removal of `Private`, which I think is universal desire.

The second is the use of `$injector` to make it clear this is Angular stuff, so allow the code to be minified without duplicating all the dependencies, and I guess tangentially to make the code still look like the `Private` code. This change is very much up for debate, and was just an idea that was thrown out when discussing the Angular styleguide (which is also kind of in flux). Sounds like folks aren't too keen on this idea, so maybe it's best to remove it for now and discuss it outside this PR.

> I think the idea is to use $injector only for nameless functions, not for things already registered with angular. 

Actually, `$injector` only works for named, registered Angular code. It doesn't work the same way the `Private` does.

Stacey-Gammon replied:
Okay folks, I reverted to the old variation.  Discussed with @cjcenizal and @w33ble and decided to move the debate around the new $injector style to a separate issue and pull out of here, in order to stay consistent with the current code, until the new pattern is embraced as a whole by the team.

Stacey-Gammon commented:
@weltenwort and @kobelb should be all good to go for a final look, thanks!
